test/wasm2c/tail-calls.txt

```diff
@@ -0,0 +1,858 @@
+  funcref_table_init(&instance->w2c_tab, elem_segment_exprs_w2c_test_e0, 1, 0u, 0, 1, instance);
+}
+
+static void init_elem_instances(w2c_test *instance) {
+}
+
+/* export: 'tailcaller' */
+void w2c_test_tailcaller(w2c_test* instance) {
+  w2c_test_tailcaller_0(instance);
+}
+
+/* export for tail-call of 'tailcaller' */
+void wasm2c_tailcall_w2c_test_tailcaller(void **instance_ptr, void *tail_call_stack, wasm_rt_tailcallee_t *next) {
+  wasm2c_tailcall_w2c_test_tailcaller_0(instance_ptr, tail_call_stack, next);
+}
+
+/* handler for missing tail-call on import: 'spectest' 'print_i32_f32' */
+#ifdef _MSC_VER
+#pragma comment(linker, "/alternatename:wasm2c_tailcall_w2c_spectest_print_i32_f32=wasm2c_fallback_test_w2c_spectest_print_i32_f32")
+void wasm2c_fallback_test_w2c_spectest_print_i32_f32(void **instance_ptr, void *tail_call_stack, wasm_rt_tailcallee_t *next)
+#else
+__attribute__((weak)) void wasm2c_tailcall_w2c_spectest_print_i32_f32(void **instance_ptr, void *tail_call_stack, wasm_rt_tailcallee_t *next)
+#endif
+{
+  next->fn = NULL;
+  struct wasm_multi_if *params = tail_call_stack;
+  w2c_spectest_print_i32_f32(*instance_ptr, params->i0, params->f1);
+}
+
+static void init_instance_import(w2c_test* instance, struct w2c_spectest* w2c_spectest_instance) {
+  instance->w2c_spectest_instance = w2c_spectest_instance;
+}
+
+void wasm2c_test_instantiate(w2c_test* instance, struct w2c_spectest* w2c_spectest_instance) {
+  assert(wasm_rt_is_initialized());
+  init_instance_import(instance, w2c_spectest_instance);
+  init_tables(instance);
+  init_elem_instances(instance);
+}
+
+void wasm2c_test_free(w2c_test* instance) {
+  wasm_rt_free_funcref_table(&instance->w2c_tab);
+}
+
+wasm_rt_func_type_t wasm2c_test_get_func_type(uint32_t param_count, uint32_t result_count, ...) {
+  va_list args;
+  
+  if (param_count == 2 && result_count == 0) {
+    va_start(args, result_count);
+    if (true && va_arg(args, wasm_rt_type_t) == WASM_RT_I32 && va_arg(args, wasm_rt_type_t) == WASM_RT_F32) {
+      va_end(args);
+      return w2c_test_i32_f32;
+    }
+    va_end(args);
+  }
+  
+  if (param_count == 0 && result_count == 0) {
+    va_start(args, result_count);
+    if (true) {
+      va_end(args);
+      return w2c_test_t1;
+    }
+    va_end(args);
+  }
+  
+  return NULL;
+}
+
+void w2c_test_tailcaller_0(w2c_test* instance) {
+  FUNC_PROLOGUE;
+  u32 var_i0, var_i2;
+  f32 var_f1;
+  var_i0 = 1u;
+  var_f1 = 2;
+  var_i2 = 0u;
+  wasm_static_assert(sizeof(struct wasm_multi_if) <= 1024);
+  CHECK_CALL_INDIRECT(instance->w2c_tab, w2c_test_i32_f32, var_i2);
+  if (!instance->w2c_tab.data[var_i2].func_tailcallee.fn) {
+    CALL_INDIRECT(instance->w2c_tab, void (*)(void*, u32, f32), w2c_test_i32_f32, var_i2, instance->w2c_tab.data[var_i2].module_instance, var_i0, var_f1);
+    goto var_Bfunc;
+  }
+  {
+    void *instance_ptr_storage;
+    void **instance_ptr = &instance_ptr_storage;
+    char tail_call_stack[1024];
```

Comment:
I wonder if this is going to put a lot of stack pressure if we have a tail called function whose body starts another tail call chain. Could we declare the tail_call_stack directly as a union of the param types (`struct wasm_multi_if`) and any result type of the callee?

Reply:
In a **pure** chain of tail calls, the instance_ptr_storage and tail_call_stack only get declared once. A tail-callee won't have to declare these when making a tail-call. However, you're right that if a tail-callee makes a normal call, and then _that_ function makes a tail-call, then another tail_call_stack will get declared as another 1024-byte array.

Unfortunately I don't think we can be a lot smarter in picking an adaptive size for the tail_call_stack. The problem is that we don't know the union of param types -- e.g. if we tail-call a function in a different module that wants to tail-call another function with a type that we've never seen.

Reply:
ah, i see. Ok, closing this as I don't see an immediate way to optimize this even in the single module scenario without a bunch of additional tracking.